include/zephyr/drivers/crc.h

```diff
@@ -0,0 +1,191 @@
+	CRC_STATE_IDLE,
+	CRC_STATE_IN_PROGRESS,
+	CRC_STATE_COMPLETE
+};
+
+/**
+ * @brief Runtime driver configuration
+ * @param mutex Zephyr mutex object to make driver thread-safe when driver
+ *		is used within multiple threads
+ */
+struct crc_data {
+	struct k_mutex mutex;
+	k_timeout_t timeout;
+};
+
+typedef uint64_t crc_polynomial_t;
+typedef uint64_t crc_result_t;
+
+/**
+ * @brief CRC context structure
+ */
+struct crc_ctx {
+	enum crc_type type;          /**< Type of CRC algorithm */
+	enum crc_state state;        /**< State of CRC calculation */
+	const void *initial_value;   /**< Initial value for CRC calculation */
+	uint32_t flags;              /**< Flags for CRC calculation */
+	crc_polynomial_t polynomial; /**< Polynomial used in CRC calculation */
+	crc_result_t result;         /**< Store the CRC result */
+};
+
+typedef int (*crc_api_lock)(const struct device *dev, k_timeout_t timeout);
+typedef int (*crc_api_unlock)(const struct device *dev);
+typedef int (*crc_api_begin)(const struct device *dev, struct crc_ctx *ctx, enum crc_type type,
+			     uint32_t flags, crc_polynomial_t polynomial,
+			     const void *initial_value);
+typedef int (*crc_api_update)(const struct device *dev, struct crc_ctx *ctx, const void *buffer,
+			      size_t bufsize);
+typedef int (*crc_api_verify)(const struct device *dev, struct crc_ctx *ctx, crc_result_t expected);
+
+/**
+ * @brief CRC driver API structure
+ */
+__subsystem struct crc_driver_api {
+	crc_api_lock crc_lock;
+	crc_api_unlock crc_unlock;
+	crc_api_begin crc_begin;
+	crc_api_update crc_update;
+	crc_api_verify crc_verify;
+};
+
+/**
+ * @brief Attempt to acquire CRC mutex lock
+ * @param dev Pointer to the device structure
+ * @param timeout Time in milliseconds to wait, or K_NO_WAIT / K_FOREVER
+ * @retval 0 if successful, negative errno code on failure
+ */
+static inline int crc_lock(const struct device *dev, k_timeout_t timeout)
+{
+	const struct crc_driver_api *api = (const struct crc_driver_api *)dev->api;
+	return api->crc_lock(dev, timeout);
+}
+
+/**
+ * @brief Unlock the CRC device
+ * @param dev Pointer to the device structure
+ * @retval 0 if successful, negative errno code on failure
+ */
+static inline int crc_unlock(const struct device *dev)
+{
+	const struct crc_driver_api *api = (const struct crc_driver_api *)dev->api;
+	return api->crc_unlock(dev);
+}
+
+/**
+ * @brief  Configure CRC unit for calculation
+ * @param dev Pointer to the device structure
+ * @param ctx Pointer to the CRC context structure
+ * @retval 0 if successful, negative errno code on failure
+ */
+static inline int crc_begin(const struct device *dev, struct crc_ctx *ctx, enum crc_type type,
+			    uint32_t flags, crc_polynomial_t polynomial, const void *initial_value)
+{
+	const struct crc_driver_api *api = (const struct crc_driver_api *)dev->api;
+	ctx->type = type;
+	ctx->flags = flags;
+	ctx->polynomial = polynomial;
+	ctx->initial_value = initial_value;
+	ctx->state = CRC_STATE_IN_PROGRESS;
```

Comment:
What if `ctx->state == CRC_STATE_IN_PROGRESS`?

Also, in this case, it would be wiser to have user fill up the structure, and only accept a `struct crc_ctx *ctx` as parameter, since it already contains all data.